#!/usr/bin/env python
PACKAGE = "frenet_optimal_planner"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add("output_path_max_size", int_t, 0, "Default: 20", 20, 5, 20)
gen.add("output_path_min_size", int_t, 0, "Default: 10", 10, 5, 20)
gen.add("soft_safety_margin", double_t, 0, "Default: 0.8", 0.8, 0.3, 1.0)

gen.add("PID_Kp", double_t, 1, "Default: 0.100", 0.100, 0, 1.0)
gen.add("PID_Ki", double_t, 1, "Default: 0.001", 0.001, 0, 1.0)
gen.add("PID_Kd", double_t, 1, "Default: 0.0",   0.0,   0, 1.0)
gen.add("num_wp_look_ahead", int_t, 2, "Default: 2", 2, 0, 10)
gen.add("stanley_overall_gain", double_t, 1, "Default: 1.0", 0.8, 0, 1.0)
gen.add("track_error_gain", double_t, 1, "Default: 1.0", 1.0, 0, 1.0)

# gen.add("ref_spline_length", double_t, 0, "Default: 50.0", 50.0, 10.0, 100.0)
gen.add("ref_speed", double_t, 0, "Default: 20.0[km/h]", 20.0, 5.0, 60.0)

gen.add("curr_lane_width", double_t, 0, "Default: 3.0", 3.0, 0, 5.0)
gen.add("left_lane_width", double_t, 0, "Default: 3.0", 3.0, 0, 5.0)
gen.add("right_lane_width", double_t, 0, "Default: 3.0", 3.0, 0, 5.0)
gen.add("center_offset", double_t, 0, "Default: 0", 0, -1.0, 1.0)
gen.add("delta_width", double_t, 0, "Default: 0.3", 1.0, 0.1, 1.0)

gen.add("max_t", double_t, 0, "Default: 5.0", 5.0, 0, 10.0)
gen.add("min_t", double_t, 0, "Default: 3.0", 3.0, 0, 10.0)
gen.add("delta_t", double_t, 0, "Default: 1.0", 1.0, 1.0, 5.0)
gen.add("tick_t", double_t, 0, "Default: 0.3", 0.3, 0, 1.0)

gen.add("target_speed", double_t, 0, "Default: 3.0", 3.0, 0, 5.0)
gen.add("delta_speed", double_t, 0, "Default: 1.0", 1.5, 0.1, 5.0)
gen.add("num_speed_sample", int_t, 0, "Default: 1", 1, 1, 5)

gen.add("k_jerk", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
gen.add("k_time", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
gen.add("k_diff", double_t, 0, "Default: 1.0", 1.0, 0, 5.0)
gen.add("k_lateral", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
gen.add("k_longitudinal", double_t, 0, "Default: 1.0", 1.0, 0, 1.0)
gen.add("k_obstacle", double_t, 0, "Default: 0.5", 0.5, 0, 1.0)

gen.add("turn_yaw_thresh", double_t, 0, "Default: 0.785", 0.785, 0, 3.14)

gen.add("regenerate_brake_threshold", double_t, 0, "Default: 0.3", 0.3, 0, 1.0)

exit(gen.generate(PACKAGE, "frenet_optimal_planner", "frenet_optimal_planner_"))
